core/embed/trezorhal/optiga/optiga.c

```diff
@@ -376,86 +440,87 @@ static int optiga_pin_stretch_secret(OPTIGA_UI_PROGRESS ui_progress,
   //   cmac_i = CMAC(optiga_cmac_key, result_i)
   //   hmac_i = HMAC(optiga_hmac_key, result_i)
   //   ecdh_i = ECDH(optiga_ecdh_key, result_i)
   //   result_{i+1} = HMAC-SHA256(secret, cmac_i || hmac_i || ecdh_i)
-  // secret = result_{PIN_STRETCH_ITERATIONS}
+  // secret = result_{PIN_STRETCH_ITERATIONS_V4}
 
   HMAC_SHA256_CTX ctx = {0};
 
   uint8_t result[OPTIGA_PIN_SECRET_SIZE] = {0};
   memcpy(result, secret, sizeof(result));
 
-  uint8_t buffer[ENCRYPT_SYM_PREFIX_SIZE + OPTIGA_PIN_SECRET_SIZE] = {0};
-  size_t size = 0;
-  for (int i = 0; i < PIN_STRETCH_ITERATIONS; ++i) {
+  for (int i = 0; i < PIN_STRETCH_ITERATIONS_V4; ++i) {
     hmac_sha256_Init(&ctx, secret, OPTIGA_PIN_SECRET_SIZE);
 
-    // Combine intermediate result with OID_PIN_CMAC.
     optiga_result res =
-        optiga_encrypt_sym(OPTIGA_SYM_MODE_CMAC, OID_PIN_CMAC, result,
-                           sizeof(result), buffer, sizeof(buffer), &size);
+        optiga_pin_stretch_common(ui_progress, &ctx, result, true);
     if (res != OPTIGA_SUCCESS) {
-      memzero(buffer, sizeof(buffer));
       memzero(result, sizeof(result));
       memzero(&ctx, sizeof(ctx));
       return res;
     }
 
-    hmac_sha256_Update(&ctx, buffer, size);
+    hmac_sha256_Final(&ctx, result);
+  }
 
-    // Combine intermediate result with OID_PIN_HMAC
-    res = optiga_encrypt_sym(OPTIGA_SYM_MODE_HMAC_SHA256, OID_PIN_HMAC, result,
-                             sizeof(result), buffer, sizeof(buffer), &size);
-    if (res != OPTIGA_SUCCESS) {
-      memzero(buffer, sizeof(buffer));
-      memzero(result, sizeof(result));
-      memzero(&ctx, sizeof(ctx));
-      return res;
-    }
+  memcpy(secret, result, sizeof(result));
+  memzero(result, sizeof(result));
+  memzero(&ctx, sizeof(ctx));
+  return OPTIGA_SUCCESS;
+}
+
+static int optiga_pin_stretch_secret(
+    OPTIGA_UI_PROGRESS ui_progress,
+    uint8_t stretched_pin[OPTIGA_PIN_SECRET_SIZE]) {
+  // This step hardens the PIN verification process in case an attacker is able
+  // to extract the secret value of a data object in Optiga that has a
+  // particular configuration, but does not allow secret extraction for other
+  // kinds of data objects. An attacker would need to be able to extract each of
+  // the secrets in the different data objects to conduct an offline brute-force
+  // search for the PIN. Thus it reduces the number of PIN values that the
+  // attacker can test in a unit of time by forcing them to involve the Optiga
+  // in each attempt, and restricts the overall number of attempts using
+  // OID_PIN_TOTAL_CTR.
 
-    hmac_sha256_Update(&ctx, buffer, size);
+  // Pseudocode for the stretching process:
+  // for i in range(PIN_STRETCH_ITERATIONS):
+  //   digest = HMAC(secret, "")
+  //   cmac_i = CMAC(optiga_cmac_key, digest)
+  //   hmac_i = HMAC(optiga_hmac_key, digest)
+  //   ecdh_i = ECDH(optiga_ecdh_key, digest)
+  //   secret = HMAC-SHA256(secret, cmac_i || hmac_i || ecdh_i)
```

Comment:
What is the reason for indexing `cmac`, `hmac` and `ecdh` but not `digest` and `secret`.

I prefer not to use indices neither here nor in `optiga_pin_stretch_secret_v4`.

Reply:
d92f0efd52fe1d4fa82981c8a03ed20b114e0df9